src/contracts/AbstractARM.sol

```diff
@@ -0,0 +1,693 @@
+        amountIn = ((amountOut * PRICE_SCALE) / price) + 1; // +1 to always round in our favor
+
+        // Transfer the input tokens from the caller to this ARM contract
+        _transferAssetFrom(address(inToken), msg.sender, address(this), amountIn);
+
+        // Transfer the output tokens to the recipient
+        _transferAsset(address(outToken), to, amountOut);
+    }
+
+    /**
+     * @notice Set exchange rates from an operator account from the ARM's perspective.
+     * If token 0 is WETH and token 1 is stETH, then both prices will be set using the stETH/WETH price.
+     * @param buyT1 The price the ARM buys Token 1 (stETH) from the Trader, denominated in Token 0 (WETH), scaled to 36 decimals.
+     * From the Trader's perspective, this is the sell price.
+     * @param sellT1 The price the ARM sells Token 1 (stETH) to the Trader, denominated in Token 0 (WETH), scaled to 36 decimals.
+     * From the Trader's perspective, this is the buy price.
+     */
+    function setPrices(uint256 buyT1, uint256 sellT1) external onlyOperatorOrOwner {
+        // Limit funds and loss when called by the Operator
+        if (msg.sender == operator) {
+            require(sellT1 >= PRICE_SCALE - MAX_PRICE_DEVIATION, "ARM: sell price too low");
+            require(buyT1 <= PRICE_SCALE + MAX_PRICE_DEVIATION, "ARM: buy price too high");
+        }
+        _setTraderates(
+            PRICE_SCALE * PRICE_SCALE / sellT1, // base (t0) -> token (t1)
+            buyT1 // token (t1) -> base (t0)
+        );
+    }
+
+    function _setTraderates(uint256 _baseToTokenRate, uint256 _tokenToBaseRate) internal {
+        require((PRICE_SCALE * PRICE_SCALE / (_baseToTokenRate)) > _tokenToBaseRate, "ARM: Price cross");
+        traderate0 = _baseToTokenRate;
+        traderate1 = _tokenToBaseRate;
+
+        emit TraderateChanged(_baseToTokenRate, _tokenToBaseRate);
+    }
+
+    ////////////////////////////////////////////////////
+    ///         Liquidity Provider Functions
+    ////////////////////////////////////////////////////
+
+    /// @notice Preview the amount of shares that would be minted for a given amount of assets
+    /// @param assets The amount of liquidity assets to deposit
+    /// @return shares The amount of shares that would be minted
+    function previewDeposit(uint256 assets) external view returns (uint256 shares) {
+        shares = convertToShares(assets);
+    }
+
+    /// @notice deposit liquidity assets in exchange for liquidity provider (LP) shares.
+    /// The caller needs to have approved the contract to transfer the assets.
+    /// @param assets The amount of liquidity assets to deposit
+    /// @return shares The amount of shares that were minted
+    function deposit(uint256 assets) external returns (uint256 shares) {
+        // Accrue any performance fees based on the increase in available assets before
+        // the liquidity asset from the deposit is transferred into the ARM
+        _accruePerformanceFee();
```

Comment:
🟠 Get ready for a real mind bender.

Consider the following:

```solidity
function previewDeposit(uint256 assets) ... {
    shares = convertToShares(assets);
}

function deposit(uint256 assets) ... {
    _accruePerformanceFee();
    shares = convertToShares(assets);
    ...
}
```

The number of shares returned by `previewDeposit()` should match the number of shares minted by `deposit()`.

So as a consequence - and hold onto your hats - one of the following must be true:

1. The shares don't match, and we have bug
2. The shares do match, and we don't even need to have _accruePerformanceFee() at all!

My mind is blown.

This also would clean up the code a lot because we currently have two separate code paths through multiple functions for computing this kind of thing, and this would collapse it down to just one. Way less chance for a bug, and guaranteed consistent behavior.

Reply:
This lets us totally remove `feesAccrued`. Withdrawing fees calculates the fees on the contract. This calculation can probably be a public view method.

Reply:
That's something I've figured out this morning. When I saw your [previous comment](https://github.com/OriginProtocol/arm-oeth/pull/13#discussion_r1783368591).
Even before adding it to the invariant, I've started to write a comment saying "This cannot match" as we are calculating fees before. But then I implemented it, and in more than 1 million calls, this never reverted! 

https://github.com/OriginProtocol/arm-oeth/blob/0a4518a4c9934ed9bd6569a8e2c0074ac50c564a/test/invariants/handlers/LpHandler.sol#L77-L79

This means that, on both options, we only have one left
> ~The shares don't match, and we have bug~
> The shares do match, and we don't even need to have _accruePerformanceFee() at all!

Apologies, I should have figured it out earlier, but I've missed out sorry.

Reply:
`setFee` would just need to call `collectFees`().

Reply:
This is a very clever idea. Nice work @DanielVF 
I've done the changes in PR https://github.com/OriginProtocol/arm-oeth/pull/30